lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -596,17 +596,56 @@ std::string SerializedModuleBaseName::getName(file_types::ID fileTy) const {
   return std::string(result.str());
 }
 
+llvm::Optional<std::string> SerializedModuleLoaderBase::getPackageInterfacePathIfInSamePackage(llvm::vfs::FileSystem &fs, ASTContext &ctx) {
+  std::string packagePath{
+    getName(file_types::TY_PackageSwiftModuleInterfaceFile)};
+
+  if (fs.exists(packagePath)) {
+    // Read the interface file and extract its package-name argument value
+    StringRef result;
+    if (auto packageFile = llvm::MemoryBuffer::getFile(packagePath)) {
+      llvm::BumpPtrAllocator alloc;
+      llvm::StringSaver argSaver(alloc);
+      SmallVector<const char*, 8> args;
+      (void)extractCompilerFlagsFromInterface(packagePath,
+                                              (*packageFile)->getBuffer(), argSaver, args);
+      for (unsigned I = 0, N = args.size(); I + 1 < N; I++) {
+        StringRef current(args[I]), next(args[I + 1]);
+        if (current == "-package-name") {
+          result = next;
+          break;
```

Comment:
It may be better not to `break` here and just keep looping. In case of flag duplication we usually pick the last one.

I would expect that when calling the compiler with `swift-frontend -package-name A -package-name B` we would already preserve B. That's from the default logic parsing options, we should try to reproduce it here. It's not applied much in the wild but it's pretty handy in tests with all the lit macros.

Reply:
will do